exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+# Module A3 - Traits and generics
+
+
+## A3.2 Local Storage Vec
+In this exercise, we'll create a type called `LocalStorageVec`, which is generic list of items that resides either on the stack or the heap, depending on its size. If its size is small enough for items to be put on the stack, the `LocalStorageVec` buffer is backed by an array. `LocalStorageVec` is not only generic over the type  (`T`) of items in the list, but also by the size (`N`) of this stack-located array using a relatively new feature called 'const generics'. Once the `LocalStorageVec` contains more items than fit in the array, a heap based [`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html) is allocated as space for the items to reside in.
+
+**Questions**
+1. When is such a data structure more efficient than a standard `Vec`?
+2. What are the downsides, compared to `Vec`?
+
+Open the `exercises/A3/2-local-storage-vec` crate. It contains a `src/lib.rs` file, meaning this crate is a library. `lib.rs` contains a number of tests, which can be run by calling `cargo test`. Don't worry if they don't pass or even compile right now: it's your job to fix that in this exercise. Most of the tests are commented out right now, to enable a step-by-step approach. **Before you begin, have a look at the code and the comments in there, they contain various helpful clues.**
+
+### A3.2.A Defining the type
+Currently, the `LocalStorageVec` `enum` is incomplete. Give it two variants: `Stack` and `Heap`. `Stack` contains two named fields, `buf` and `usize`. `buf` will be the array containing `N` items of type `T`. The `Heap` variant has an unnamed field containing a `Vec<T>`. If you've defined the `LocalStorageVec` variants correctly, running `cargo test` should output something like
```

Comment:
I understand the desire to have students create their own enum, but consider just giving it to them: if they get stuck here that is a blocker for all the next exercises, and the specification does not give them any real freedom in what the type must look like in any case.